### PR TITLE
Add typings for Figma Plugin API

### PR DESCRIPTION
**Pull request by rudi-c**
This PR adds typings for the the Figma Plugin API that was released last week (I work at Figma).

Note that this is not a typings file for a traditional NPM package with source code on Github. These are the typings file for an API exposed to plugin code when they run inside a special sandbox inside the Figma design tool, on our website.

Note also that there already exists an NPM package with the name `figma`. That package is currently empty and owned by the CTO of Figma @evanw, so that's not a problem even though the two packages are unrelated.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/figma/tslint.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "extends": "dtslint/dt.json" ,
+    "rules": {
+        "dt-header": false,
+        "npm-naming": false,
+        "no-any-union": false,
```
Comment (sandersn) [inline types/figma/tslint.json]: New packages should not override any lint rules. You explained why dt-header and npm-naming don't apply (thanks! it was a very clear explanation), but please remove the other overrides.

File: types/figma/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,666 @@
...
+
+    createImage(data: Uint8Array): Image
+    getImageByHash(hash: string): Image
+
+    group(nodes: ReadonlyArray<BaseNode>, parent: BaseNode & ChildrenMixin, index?: number): FrameNode
+    flatten(nodes: ReadonlyArray<BaseNode>, parent?: BaseNode & ChildrenMixin, index?: number): VectorNode
+}
+
+interface ClientStorageAPI {
+    getAsync(key: string): Promise<any | undefined>
```
Comment (sandersn) [inline types/figma/index.d.ts]: `unknown` is better than `any | undefined` here, unless you want to explicitly exclude `null`, in which case `{} | undefined` is the shortest type you could write. But `object | string | number | boolean | symbol | bigint | undefined` might be easier to read.

**Review by sandersn — CHANGES_REQUESTED** [review]
Need to remove some lint rule overrides.

**Comment by chrisdrackett** [comment]
@rudi-c do these types cover the Figma Rest API as well? If not, are their plans to officially publish these? I was thinking of submitting https://github.com/jongold/figma-js/blob/master/src/figmaTypes.ts to DefinitelyTyped but will hold off if there are official plans!

**Comment by rudi-c** [comment]
Ah my bad, I made the changes but then pushed the wrong branch. Re-opened here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/37748

@chrisdrackett No, not sure if it makes sense to have both the plugin API types and the REST API types in the same package. It doesn't seem like anyone on the infra team is currently planning on publishing types for the API.

**Pull request closed without merging**